### PR TITLE
Refactors code to use u16string instead of wstring.

### PR DESCRIPTION
**Pull request by lexicalunit**
Resolves issue #86.

### CODE REVIEW

**Comment by lexicalunit** [comment]
This PR switches everything over to `u16string` and strips out `wstring` and `wchar_t` everywhere. It should make unicode work on *nix builds now and leave Windows builds unaffected. There's a lot of changes here though. For example the `swprintf()`, `wcsftime()`, `wcslen()`, `wcsncmp()` family of functions are completely gone now. Instead, we narrow or widen strings as needed and do any manipulation with `string` rather than `u16string`. String streams are removed as they're not implemented for `u16string`.

Would love to get your thoughts and tests on a Windows system @mloskot.

**Comment by mloskot** [comment]
On Windows, the new implementation hits a bug in VS2015 (including Update 1):

```
nanodbc.lib(nanodbc.obj) : error LNK2001: unresolved external symbol
"__declspec(dllimport) public: static class std::locale::id std::codecvt<char16_t,char,struct _Mbstatet>::id"
(__imp_?id@?$codecvt@_SDU_Mbstatet@@@std@@2V0locale@2@A)
```

The error is caused by the line like this (`nanodbc.cpp` and `example_unicode_utils.h`):
```
std::wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t>().from_bytes(in);
```

In discussion here, [VS 2015 RC linker std::codecvt error](https://social.msdn.microsoft.com/Forums/en-US/8f40dcd8-c67f-4eba-9134-a19b9178e481/vs-2015-rc-linker-stdcodecvt-error), Stephan T. Lavavej confirms the bug is still in the VS2015 Update 1, and recommends:

> Use wchar_t or unsigned short instead of char16_t, and it should work.

**Comment by mloskot** [comment]
I have tested the workaround #107 with SQL Server and it seems working fine. I hope I haven't missed any additional hacks :)

**Comment by lexicalunit** [comment]
I broke this PR up into 9 individual sequential commits, each one pertaining to one logical change in the code. It should make the changes easier to reason about :)

**Comment by mloskot** [comment]
:+1:

**Comment by lexicalunit** [comment]
Yay, merging!

File: src/nanodbc.cpp [hunk 1]
```diff
@@ -2596,81 +2595,107 @@ inline void result::result_impl::get_ref_impl<string_type>(short column, string_
...
-                , column_size + 2
-                , NANODBC_TEXT("%.*lf")         // restrict the number of digits
+            std::string buffer;
+            const SQLULEN width = column_size + 2; // account for decimal mark and sign
+            buffer.reserve(width + 1); // ensure terminating null
+            buffer.resize(buffer.capacity());
+            using std::fill;
+            fill(buffer.begin(), buffer.end(), '\0');
+            const double data = *reinterpret_cast<double*>(col.pdata_ + rowset_position_ * col.clen_);
+            printf("data: %.*lf\n", col.scale_, data);
```
Comment (mloskot) [inline src/nanodbc.cpp]: It looks like printf-based debugging left over, doesn't it?

Comment (lexicalunit) [inline src/nanodbc.cpp]: Yep, will remove.

**Pull request merged**